File: Unstop/100DayCode/Day 3/mocha'sAlienDict.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

bool wordbreak(string s, unordered_set<string>& worddict) {
    if (worddict.empty()) return false;
    vector<bool> dp(s.size() + 1, false);
    dp[0] = true;

    for (int i = 1; i <= s.size(); i++) {
        for (int j = i - 1; j >= 0; j--) {
            if (dp[j]) {
                string word = s.substr(j, i - j);
                if (worddict.find(word) != worddict.end()) {
                    dp[i] = true;
                    break;
                }
            }
        }
    }
    return dp[s.size()];
}

int main() {
    string s;
    cin >> s;
    int n;
    cin >> n;
    unordered_set<string> worddict;
    for (int i = 0; i < n; i++) {
        string str;
        cin >> str;
        worddict.insert(str);
    }
    wordbreak(s, worddict) ? cout << "true" : cout << "false";
    return 0;
}

File: Unstop/100DayCode/Day 3/mocha'sAlienDict.py
def wordbreak(s, worddict):
    if not worddict:
        return False
    wordset = set(worddict)
    dp = [False] * (len(s) + 1)
    dp[0] = True

    for i in range(1, len(s) + 1):
        for j in range(i - 1, -1, -1):
            if dp[j]:
                word = s[j:i]
                if word in wordset:
                    dp[i] = True
                    break

    return dp[-1]

if __name__ == "__main__":
    s = input().strip()
    n = int(input())
    worddict = []
    for _ in range(n):
        worddict.append(input().strip())
    print("true" if wordbreak(s, worddict) else "false")
